gms/gossiper.cc

```diff
@@ -557,6 +581,50 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
     }
 }
 
+// Runs inside seastar::async context
+void gossiper::apply_state_locally_without_listener_notification(std::unordered_map<inet_address, endpoint_state> map) {
```

Comment:
Does this not duplicate apply_state_locally()? Let's merge the implementations.

Reply:
We can merge them. The downside is we will add a lot of if branches which makes apply_state_locally even harder to understand.

Reply:
I'm sure it can be abstracted away. For example, a helper function `notify = [should_notify] (...) { if (should_notify) { ... } };` can make the branches invisible.